embassy-extras/src/mpsc.rs

```diff
@@ -0,0 +1,257 @@
+//!
+//! This module provide a bounded channel that has a limit on the number of
+//! messages that it can store, and if this limit is reached, trying to send
+//! another message will wait until a message is received from the channel.
+//!
+//! Similar to the `mpsc` channels provided by `std`, the channel constructor
+//! functions provide separate send and receive handles, [`Sender`] and
+//! [`Receiver`]. If there is no message to read, the current task will be
+//! notified when a new value is sent. [`Sender`] allows sending values into
+//! the channel. If the bounded channel is at capacity, the send is rejected
+//!
+//! # Disconnection
+//!
+//! When all [`Sender`] handles have been dropped, it is no longer
+//! possible to send values into the channel. This is considered the termination
+//! event of the stream.
+//!
+//! If the [`Receiver`] handle is dropped, then messages can no longer
+//! be read out of the channel. In this case, all further attempts to send will
+//! result in an error.
+//!
+//! # Clean Shutdown
+//!
+//! When the [`Receiver`] is dropped, it is possible for unprocessed messages to
+//! remain in the channel. Instead, it is usually desirable to perform a "clean"
+//! shutdown. To do this, the receiver first calls `close`, which will prevent
+//! any further messages to be sent into the channel. Then, the receiver
+//! consumes the channel to completion, at which point the receiver can be
+//! dropped.
+//!
+//! This channel and its associated types were derived from https://docs.rs/tokio/0.1.22/tokio/sync/mpsc/fn.channel.html
+
+use core::fmt;
+use core::marker::PhantomData;
+use core::mem;
+
+/// Send values to the associated `Receiver`.
+///
+/// Instances are created by the [`channel`](channel) function.
+pub struct Sender<T> {
+    _phantom_data: PhantomData<T>,
+}
+
+/// Receive values from the associated `Sender`.
+///
+/// Instances are created by the [`channel`](channel) function.
+pub struct Receiver<T> {
+    _phantom_data: PhantomData<T>,
+}
+
+/// Creates a bounded mpsc channel for communicating between asynchronous tasks
+/// with backpressure.
+///
+/// The channel will buffer up to the provided number of messages.  Once the
+/// buffer is full, attempts to `send` new messages will wait until a message is
+/// received from the channel. The provided buffer capacity must be at least 1.
+///
+/// All data sent on `Sender` will become available on `Receiver` in the same
+/// order as it was sent.
+///
+/// The `Sender` can be cloned to `send` to the same channel from multiple code
+/// locations. Only one `Receiver` is supported.
+///
+/// If the `Receiver` is disconnected while trying to `send`, the `send` method
+/// will return a `SendError`. Similarly, if `Sender` is disconnected while
+/// trying to `recv`, the `recv` method will return a `RecvError`.
+///
+/// # Panics
+///
+/// Panics if the buffer capacity is 0.
+pub fn channel<'a, T>(buf: &'a mut [u8]) -> (Sender<T>, Receiver<T>) {
```

Comment:
I don't think this is possible to implement without `alloc`. `Sender` and `Receiver` would need a lifetime param so they can borrow the buffer (`Sender<'a, T>`). 

We probably need to store more state than just the buf (wakers, fill amount..) too, so we probably need a Channel type to hold it (and have Sender/Receiver borrow the Channel instead of the buf)

Reply:
Absolutely. I’ve not fleshed out the structs at all yet.

Reply:
Btw the approach to construction and retention of this buffer slice will be as be the existing circular buffer we have.

Reply:
> I don't think this is possible to implement without `alloc`.

I think I've now caught up with you. :-)

The reasons for `alloc`, so far as I can see are so that Sender objects can be cloned between tasks. Is this what you were thinking @Dirbaio?

Reply:
No longer using the circular buffer.